spec/libraries/course/conditional/conditional_dependency_graph_spec.rb

```diff
@@ -0,0 +1,229 @@
+
+    @g = DummyConditionalCondition.build([], 'G')
+    @h = DummyConditionalCondition.build([@g], 'H')
+    @i = DummyConditionalCondition.build([@h], 'I')
+
+    [@b, @d, @e, @h, @g, @a, @i, @c, @f]
+  end
+
+  describe '.build' do
+    subject { Course::Conditional::ConditionalDependencyGraph }
+
+    context 'when dependencies do not contain cycle' do
+      before(:each) do
+        @simple_graph = create_simple_graph
+      end
+
+      it 'build dependency graph' do
+        expect(subject.build(@simple_graph)).to be_instance_of(
+                                                 Course::Conditional::ConditionalDependencyGraph)
+      end
+    end
+
+    context 'when dependencies contains cycle' do
+      before(:each) do
+        @cyclic_graph = create_cyclic_graph
+      end
+
+      it 'raise ArgumentError' do
+        expect{ subject.build(@cyclic_graph) }.to raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe '#resolve' do
+    context 'simple graph' do
+      before(:each) do
+        @simple_graph = create_simple_graph
+        allow_any_instance_of(DummyConditionalCondition)
+          .to receive(:satisfied_by?).and_return(false)
+      end
+
+      subject { Course::Conditional::ConditionalDependencyGraph.build(@simple_graph) }
+
+      context 'when resolved conditions do not resolve other conditions' do
+        context 'when no conditions are satisfied initially' do
+          it 'returns only conditionals with no condition' do
+            expect(subject.resolve([], double)).to contain_exactly(@a, @b)
+          end
+        end
+
+        context 'when condition A & B are satisfied initially' do
+          it 'returns only conditionals that have been resolved' do
+            expect(subject.resolve([@a, @b], double)).to contain_exactly(@a, @b, @c)
+          end
+        end
+
+        context 'when all conditions are satisfied initially' do
+          it 'returns all conditionals' do
+            expect(subject.resolve(@simple_graph, double)).to match_array(@simple_graph)
+          end
+        end
+      end
+
+      context 'when resolved conditionals do resolve other conditions' do
+
+        context 'when no conditions are satisfied initially' do
+          context 'when conditions for A & B are satisfied by the user' do
+            it 'returns conditionals up to C' do
+              allow(@a).to receive(:satisfied_by?).and_return(true)
+              allow(@b).to receive(:satisfied_by?).and_return(true)
+              expect(subject.resolve([], double)).to contain_exactly(@a, @b, @c)
+            end
+          end
+
+          context 'when all cascading conditions are satisfied by user'
+            it 'returns all conditionals' do
+              allow_any_instance_of(DummyConditionalCondition)
+                .to receive(:satisfied_by?).and_return(true)
+              expect(subject.resolve([], double)).to match_array(@simple_graph)
+            end
+          end
```

Comment:
`end` at 163, 10 is not aligned with `context 'when no conditions are satisfied initially' do` at 148, 8.